### DISCUSSION TITLE
How do you build forms when you have relation fields that have tens of thousands of entities to choose from?

### DISCUSSION THREAD

**Discussion by jerrac** [symfony/symfony#49569] [Q&A] [answered]
Ok, so, I've been trying to find my way through or around this issue for a few days now.  At this point I think I have a couple strategies, but I keep getting stuck. So here is what I am trying to do.

I building a Bible Reading Plan generator. The idea is you enter some criteria, and my app will generate a plan for you. Then you refine the plan by adjusting what and when things are read.

Each plan is split up into days, then sessions, then sections. Where each section contains a set of verses to read. 

So I have entities for each book, chapter, and verse in the Bible. According to my database, there are 31,103 verses in the Bible. A reading plan will need to reference all, or at least thousands, of those verses.

A plan gets split up into, days, that get split up into sessions, and the sessions are split up into sections. Plan, day, session, and section are all entity types.

What I've been trying to do is only load one day at a time when refining the plan. That should let me restrict how many verses need to be referenced. But I keep finding myself stymied by not finding a way to restrict what choices the form loads based on the data being loaded into the form.... 

The idea would be to look at the chapters referenced by the section being loaded, then only grab the verses for those chapters, and maybe a chapter or two before and after the selected chapters.

I'd think a choice_loader closure with a query builder based on the current data would work. But how do I get access to the current data in a way that lets me build the query builder?

I've read a dozen or so posts about similar issues found via searches, but nothing has clicked into place yet. I also experimented with data mappers, but it just didn't work quite right.  I tried the UX autocomplete field, but ran into the same issue of not being able to filter the custom query I tried to build for it. 

I guess that all boils down to:

A) What general strategies are there for dealing with thousands of entities in a relation field?

B) How do I get access to the data being loaded into a form while building the form being loaded?

Thanks in advance!

**Comment by lyrixx** [comment, marked as answer]
I usually use autocomplete in such situations. However it was by hand because Symfony/UX did not exist the last time I needed it
